drivers/rtc/rtc_nxp_irtc.c

```diff
@@ -0,0 +1,463 @@
+/*
+ * Copyright 2024 NXP
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT nxp_irtc
+
+#include <zephyr/devicetree.h>
+#include <zephyr/drivers/rtc.h>
+#include <zephyr/irq.h>
+
+struct nxp_irtc_config {
+	RTC_Type *base;
+	void (*irq_config_func)(const struct device *dev);
+	bool is_output_clock_enabled;
+	uint8_t clock_select;
+	uint8_t alarm_match_flag;
+};
+
+struct nxp_irtc_data {
+	bool is_dst_enabled;
+#ifdef CONFIG_RTC_ALARM
+	rtc_alarm_callback alarm_callback;
+	void *alarm_user_data;
+	uint16_t alarm_mask;
+	bool is_alarm_pending;
+#endif /* CONFIG_RTC_ALARM */
+};
+
+/* The IRTC Offset is 2112 instead of 1990 [2112 - 1990] -> [212] */
+#define RTC_NXP_IRTC_YEAR_OFFSET 212
+
+/* Utility function for testing if bit for alarm mask is enabled. */
+#define RTC_NXP_RTCTIME_COMPARE_FLAG(mask, flag) (((mask) == 0) || (((mask) & (flag)) && (flag)))
+
+/* Utility function to check the bounds of passed in rtc_time */
+static uint8_t is_rtc_time_in_bounds(const struct rtc_time *time_value, uint16_t mask)
+{
+	if ((RTC_NXP_RTCTIME_COMPARE_FLAG(mask, RTC_ALARM_TIME_MASK_SECOND)) &&
+	    (time_value->tm_sec < 0 || time_value->tm_sec > 59)) {
+		return 0;
+	}
+
+	if ((RTC_NXP_RTCTIME_COMPARE_FLAG(mask, RTC_ALARM_TIME_MASK_MINUTE)) &&
+	    (time_value->tm_min < 0 || time_value->tm_min > 59)) {
+		return 0;
+	}
+
+	if ((RTC_NXP_RTCTIME_COMPARE_FLAG(mask, RTC_ALARM_TIME_MASK_HOUR)) &&
+	    (time_value->tm_hour < 0 || time_value->tm_hour > 23)) {
+		return 0;
+	}
+
+	if ((RTC_NXP_RTCTIME_COMPARE_FLAG(mask, RTC_ALARM_TIME_MASK_MONTHDAY)) &&
+	    (time_value->tm_mday < 1 || time_value->tm_mday > 31)) {
+		return 0;
+	}
+
+	if ((RTC_NXP_RTCTIME_COMPARE_FLAG(mask, RTC_ALARM_TIME_MASK_MONTH)) &&
+	    (time_value->tm_mon < 0 || time_value->tm_mon > 11)) {
+		return 0;
+	}
+
+	if ((RTC_NXP_RTCTIME_COMPARE_FLAG(mask, RTC_ALARM_TIME_MASK_YEAR)) &&
+	    ((time_value->tm_year - RTC_NXP_IRTC_YEAR_OFFSET) < -128 ||
+	     (time_value->tm_year - RTC_NXP_IRTC_YEAR_OFFSET) > 127)) {
+		return 0;
+	}
+
+	if ((RTC_NXP_RTCTIME_COMPARE_FLAG(mask, RTC_ALARM_TIME_MASK_WEEKDAY)) &&
+	    (time_value->tm_wday < -1 || time_value->tm_wday > 6)) {
+		return 0;
+	}
+
+	if ((mask == 0) && (time_value->tm_isdst < -1 || time_value->tm_isdst > 1)) {
```

Comment:
I think you can add this to the rtc_utils.c if it's missing

Reply:
Its code that is really only valid for this driver at the moment.